[정렬] 과제 제출

Doren1 opened the pull request:
정렬 과제 제출합니다! 그런데 1758번에서 출력값은 백준과 똑같이 나오는데 틀렸다고 떠서 어느 부분에서 잘못된 건지 알려주시면 감사하겠습니다 ㅠㅠ! 

기존 제출: 1026, 1758, 11651번

Doren1 committed:
[정렬] 9월 7일

1026번, 11651번 풀었습니다!

Doren1 committed:
Create [정렬]1758번.cpp

jaeseo222 commented on [정렬]11651번.cpp:
```diff
@@ -0,0 +1,40 @@
+
+using namespace std;
+
+struct point {
+	int x;
+	int y;
+};
```
p2.

구조체를 선언해주셨네요!! x, y 처럼 직관적인 변수명은 참조할 때 헷갈리지 않는 장점이 있을 수 있죠!
이렇듯 **2개의 변수를 한 번에 효율적으로 관리**해야 할 때 구조체도 좋지만, **`pair`라는 두 개의 자료형을 하나로 묶어서 사용할 수 있는 Container도 있습니다.** 😄
`pair`은 **벡터에 자료형으로 선언**하여 2차원 배열처럼 사용하기도 하고, 현재 문제처럼 **좌표계**를 표현할 때 사용하기도 합니다. 아래 사용 방법 블로그를 첨부합니다!

📌 [pair 사용 방법](https://novlog.tistory.com/13)

두 자료형을 하나로 묶어 사용할 때 **구조체** vs **pair** 둘 중 편한 걸로 사용해주시면 되며, 보통 코테에선 변수명을 선언해주지 않아도 되는 **pair** 을 사용합니다!

🔥🔥 그럼, `pair`을 `sort` 했을 땐 두 자료형 중 어느 것을 기준으로 정렬될까요? 이를 고려하여 **비교 함수**를 구현해주시면 좋을 것 같습니다.

jaeseo222 reviewed (commented):
**P2. 11651 코드 리뷰 완료**

안녕하세요 도희님 😄😄 코드도 너무 깔끔하고 주석도 작성해주셔서 리뷰 드리기 너무 좋았습니다!

간단한 코멘트 달았으니, 확인 부탁 드립니다 !
**수정하신 부분 있으시면 리뷰어로 불러주셔도 좋습니다!** 수고 많으셨습니다.
감사합니다. 🤗🤗

jk0527 commented on [정렬]1026번.cpp:
```diff
@@ -0,0 +1,33 @@
+
+using namespace std;
+
+// ����� ������, ����� ���������� ����ϸ� �ּڰ��� �ȴٴ� ����Ͽ� ����
```
주석 아주 깔끔 명료하고 좋습니다!!

jk0527 reviewed (approved):
P3. 1026 코드 리뷰 완료

정말 저희가 준비한 샘플코드와 똑같이 구현해주셨어요! 로직도 좋고 코드도 깔끔해서 1026번은 이대로 머지하셔도 될 것 같아요!!😊😊 수고하셨습니다!! 궁금한 점이 생기면 리뷰어 호출해주세요!

bsa0322 commented on [정렬]1758번.cpp:
```diff
@@ -0,0 +1,41 @@
+#include <vector>
+
+using namespace std;
+vector<int> arr;
+
```
p2. 전역변수 사용은 최대한 지양하는 편이 좋습니다!! 지금 문제에서는 괜찮지만, 나중에 협업을 하게 될 경우 전역변수에서 선언한 변수를 지역변수로 또 선언하거나, 여러 군데에서 수정하거나, 등등.. 충돌이 날 상황이 많이 생길 수 있어요! 

물론 전역변수를 사용하는 편이 더 좋은 경우도 많습니다! 재귀함수에서 사용하는 배열일 경우엔 메모리를 효율적으로 관리하기 위해 주로 전역변수로 선언하고 사용하곤 하죠.

현재 문제에서 사용한 `arr 배열`은  함수의 **매개변수**로 넘겨서 사용하면 어떨까요?

bsa0322 commented on [정렬]1758번.cpp:
```diff
@@ -0,0 +1,41 @@
+			count++;   //�ε��� �� �ϳ� ��
+		}
+		
+	}
+	cout << tip_sum << "\n";
```
p1. 디버깅을 제외한 출력은 가급적 `main`에서 진행해주시는 편이 좋아요~! 함수는 용도를 명확하게 작성하는 것이 중요한데, 출력을 함께하게 되면 용도가 불명확해질 수 있기 때문이죠! 따라서 **연산에 해당하는 부분은 함수에서 수행하고 정답을 리턴하는 형태**로 작성해보면 어떨까요?

특히 코테에서는` solution 타입`이 거의 대부분이기 때문에, 문제의 정답을 리턴하는 형식의 함수를 짜는데 익숙해지시면 좋습니다!!

bsa0322 commented on [정렬]1758번.cpp:
```diff
@@ -0,0 +1,41 @@
+using namespace std;
+vector<int> arr;
+
+void max_tip(int left, int right) {
+	int p1 = left;  //���ʺ��� �������� ���鼭 ����ų ������
```
p3. 함수 이름은 **카멜표기법**에 맞춰서 작성 부탁드려요 🔥

bsa0322 commented on [정렬]1758번.cpp:
```diff
@@ -0,0 +1,41 @@
+	int count = 1;
+	int tip_sum = 0;
+
+	while (p1 <= right) {
+		if ((arr[p1] - (count - 1)) < 0) {  //���� ���� �ٷ� ��ǹ� ����������
+			break;
+		}
+		else {
+			tip_sum += arr[p1] - (count - 1);  //�� ���(��� ���)
+			p1++;  //��� ĭ ����Ű����
+			count++;   //�ε��� �� �ϳ� ��
+		}
```
팁이 음수면 바로 빠져나오는 형태로 짜주셨네요! 연산의 횟수를 조금이라도 줄일 수 있으니 아주 좋아요!! 🔥
아래 가독성 관련 코멘트 남겼습니다.

**p2.** 우선 p1은 항상 다음 칸을 가리키게 될 거예요. 팁이 음수일 경우엔 다음 칸을 가리킬 필요는 없지만, 어차피 반복문을 빠져나오게 될테니 큰 상관은 없겠죠! 그러므로 while문의 조건에 함께 포함해서 코드를 줄일 수 있을 것 같네요! 그리고 count 변수는 항상 p1과 같이 증가하고 있어요. 그렇다면 두 변수 중 하나만 사용하는 건 어떨까요? 

더 나아가서, 이렇게 포인터값이 계속 증가하는 반복을 수행할 경우 while문보다 더 간편하게 쓸 수 있는 반복문이 있어요! 

마지막으로, if문에 걸렸다면 반복문을 빠져나갈테고, if문에 걸리지 않았다면 자동으로 다음 줄을 실행할테니 현재 코드에서 꼭 필요하지 않은 코드가 보이네요!

bsa0322 reviewed (changes requested):
**p1. 1758 코드리뷰 완료**
도희님 안녕하세요 😊 리뷰가 늦어 죄송합니다 😭 1758 힌트 남겼으니, 수정이나 질문 있으시면 저 다시 리뷰어로 호출해주세요!! 리뷰어 호출은 오른쪽 상단에 reviewrs에 **새로고침**이모지를 눌러주시면 됩니다!
우선 문제의 틀렸습니다 먼저 해결하신 후, 남긴 코멘트들 확인해주시면 좋을 것 같습니다!! 첫 과제 수고하셨습니다 ✨

**1758 Hint.**
문제의 n과 팁의 범위를 다시 살펴볼까요? 만약 두 값 모두 100,000으로 들어오게 된다면 강호가 받을 수 있는 최대 팁의 범위는 어떻게 될까요? 마지막으로 int형의 범위를 살펴보면 해결하실 수 있을 거 같습니다!!

Doren1 committed:
[맵과 셋]

10757, 11723, 7785, 9375번 제출합니다!

Doren1 committed:
[정렬]1758번_수정

늦어서 죄송합니다ㅠㅠㅠ코드리뷰 반영했습니다!

질문이 하나 있는데, maxTip함수내에서 p1을 long long 타입으로 선언하면 안되는 걸까요?? 범위는 int로 해도 될 것 같아서 int로 선언했는데, overcasting경고가 떠서 질문드립니다 ...!

bsa0322 commented on [정렬]1758번_수정.cpp:
```diff
@@ -0,0 +1,43 @@
+
+	while ((p1 <= right) && ((arr[p1] - p1) >= 0)) { //���� ���� �ٷ� Ż��
+		
+		tip_sum += (arr[p1] - p1);  //�� ���(��� ���)
+		p1++;  //��� ĭ ����Ű����
```
증감연산자의 특성을 생각하면 14번째 줄과 15번째 줄은 한 줄로 합칠 수 있을 것 같네요~!

bsa0322 reviewed (approved):
코드 확인했습니다!! 사소한 코멘트 남겼으니 확인 부탁드려요~! 아예 반복이 도는 조건을 while문 안에 2개로 해서 구현해주신 방식 좋은 거 같아요!!!! 

다만, 가독성 부분에서 취향 차이로 for문을 사용하는 사람들도 많이 있으니 샘플코드도 한 번 확인해주시면 좋을 것 같습니다! 수고하셨습니다 ❤❤❤
